### ISSUE TITLE
[FEATURE] Provide Monitoring and or Logging as an AddOn

### ISSUE THREAD

**Issue by abonillabeeche** [harvester/harvester#2992]
**Is your feature request related to a problem? Please describe.**
Monitoring and Logging (hopefully separate) should be enabled as an AddOn vs a mandatory component. Edge and smaller Harvester deployments would benefit from a leaner cluster that may not necessarily include Logging or Monitoring.

**Describe the solution you'd like**
Ship Monitoring and Loggin as AddOns

**Additional context**
```
cattle-logging-system       harvester-default-event-tailer-0                         1m           16Mi
cattle-logging-system       rancher-logging-655578478b-pk9vc                         3m           126Mi
cattle-logging-system       rancher-logging-kube-audit-fluentbit-ctvzv               1m           38Mi
cattle-logging-system       rancher-logging-kube-audit-fluentd-0                     1m           112Mi
cattle-logging-system       rancher-logging-rke2-journald-aggregator-wxs5g           1m           37Mi
cattle-logging-system       rancher-logging-root-fluentbit-s88gp                     5m           47Mi
cattle-logging-system       rancher-logging-root-fluentd-0                           2m           131Mi
cattle-monitoring-system    alertmanager-rancher-monitoring-alertmanager-0           1m           50Mi
cattle-monitoring-system    prometheus-rancher-monitoring-prometheus-0               73m          1541Mi
cattle-monitoring-system    rancher-monitoring-grafana-8559b68898-nlhdk              3m           142Mi
cattle-monitoring-system    rancher-monitoring-kube-state-metrics-77ddfd789b-x676j   2m           66Mi
cattle-monitoring-system    rancher-monitoring-operator-bd988f5fb-jmhx7              1m           214Mi
cattle-monitoring-system    rancher-monitoring-prometheus-adapter-79d8db9697-94tcl   189m         239Mi
cattle-monitoring-system    rancher-monitoring-prometheus-node-exporter-645rl        2m           38Mi
```

**Comment by w13915984028** [comment]
UI related work:
How to manually convert `managedchart` to `addon`, and test UI:
https://github.com/harvester/harvester/issues/3235#issuecomment-1335246014

Backend source code:
The source code modification (WIP) to remove `managedchart` by default:
https://github.com/harvester/harvester-installer/commit/832e6a8e394d2c1e4d44921d40496d5607a4cd5a

**Comment by abonillabeeche** [comment]
With Monitoring/Logging Rep 0
![Screenshot 2022-12-21 at 11 12 13 AM](https://user-images.githubusercontent.com/5146963/208967139-7f092e0b-d88b-45d0-84d2-425c636fde31.png)

With Monitoring/Logging Rep 1
![Screenshot 2022-12-21 at 12 23 51 PM](https://user-images.githubusercontent.com/5146963/208967163-606d59e3-04b6-4d0b-92f3-caa6a18c18d6.png)

**Comment by guangbochen** [comment]
Move back to dev, will move this to test after the upgrade PR is merged.

**Comment by TachunLin** [comment]
### Test Plan 

#### Fresh install Harvester 

1. - [x] Single node (default) : 
      - [x]  Monitoring and logging addon are disabled by default
      - [x]  No pods exists on `cattle-logging-system` and `cattle-monitoring-system` namespace
      - [x]  (Cluster)Flow/(Cluster)Output display by default  
      
2. - [x] Multi nodes (default): 
      - [x]  Monitoring and logging addon are disabled by default
      - [x]  No pods exists on `cattle-logging-system` and `cattle-monitoring-system` namespace
      - [x]  (Cluster)Flow/(Cluster)Output display by default  

3.  - [x] Single node (enabled) :
      - [x]  Can enable monitoring and logging addon
      - [x]  Monitoring pods created and running correctly on `cattle-monitoring-system` namespace
      - [x]  Logging pods created and running correctly on `cattle-logging-system` namespace 
      - [x]  Monitoring chart display correctly on cluster and VM metrics
      - [x]  (Cluster)Flow/(Cluster)Output work as expected
      
4.  - [x] Multi nodes (enabled) : 
      - [x]  Can enable monitoring and logging addon
      - [x]  Monitoring pods created and running correctly on `cattle-monitoring-system` namespace
      - [x]  Logging pods created and running correctly on `cattle-logging-system` namespace 
      - [x]  Monitoring chart display correctly on cluster and VM metrics
      - [x]  (Cluster)Flow/(Cluster)Output work as expected

5. - [x] Single nodes (disabled): 
      - [x] Monitoring and logging addon can be disabled 
      - [x] Monitoring pods and namespace `cattle-monitoring-system` namespace are deleted 
      - [x] Logging pods and namespace `cattle-logging-system` namespace are deleted 
      - [x] Harvester system work without error 
      
6. - [x] Multiple nodes (disabled): 
      - [x] Monitoring and logging addon can be disabled 
      - [x] Monitoring pods and namespace `cattle-monitoring-system` namespace are deleted 
      - [x] Logging pods and namespace `cattle-logging-system` namespace are deleted 
      - [x] Harvester system work without error 

#### Upgrade from v1.1.2 to v1.2.0 master  

1.  - [x] Single node:
      - [x]  Monitoring and logging addons keep enabled
      - [x]  Can disable monitoring and logging addon
      - [x]  Monitoring pods created and running correctly on `cattle-monitoring-system` namespace
      - [x]  Logging pods created and running correctly on `cattle-logging-system` namespace 
      - [x]  Monitoring chart display correctly on cluster and VM metrics
      - [x]  (Cluster)Flow/(Cluster)Output work as expected
2.  - [x] Multi nodes: 
      - [x]  Monitoring and logging addons keep enabled
      - [x]  Can disable monitoring and logging addon
      - [x]  Monitoring pods created and running correctly on `cattle-monitoring-system` namespace
      - [x]  Logging pods created and running correctly on `cattle-logging-system` namespace 
      - [x]  Monitoring chart display correctly on cluster and VM metrics
      - [x]  (Cluster)Flow/(Cluster)Output work as expected

#### Additional
1. Check the resource saving and usage

**Comment by TachunLin** [comment]
* Three nodes Harvester resource usage before enabling monitoring and logging addon
![image](https://github.com/harvester/harvester/assets/29251855/41c9c292-5f10-474a-abea-8a4ad92ec88f)

* Three nodes Harvester resource usage after enabling addons
![image](https://github.com/harvester/harvester/assets/29251855/19748560-0de7-4845-9612-f2b61435b60c)
  - reserved CPU increased on each nodes  
      - node1: 0.2 core, node2: 0.62 core, node3: 1.45 core
  - used memory increased on each nodes 
      - node1: 1 GiB, node2: 1.48 GiB, node3: 2.35 GiB

* Single node Harvester resource usage before enabling monitoring and logging addon
  ![image](https://github.com/harvester/harvester/assets/29251855/7d82cae7-3622-4527-89b6-6fc95516611c)

* Single node Harvester resource usage after enabling addons
  ![image](https://github.com/harvester/harvester/assets/29251855/52a0a813-d26f-4410-a875-ff7b20bbcb56)


* Pod status before enabling addons
No pods exists on `cattle-logging-system` and `cattle-monitoring-system` namespace

* Pod status after enabling addons
```
cattle-logging-system       harvester-default-event-tailer-0                          1/1     Running     0               2m33s
cattle-logging-system       helm-install-rancher-logging-vn4jq                        0/1     Completed   0               2m43s
cattle-logging-system       rancher-logging-5868b84cf-gwc6v                           1/1     Running     0               2m37s
cattle-logging-system       rancher-logging-kube-audit-fluentbit-6q6m6                1/1     Running     0               2m9s
cattle-logging-system       rancher-logging-kube-audit-fluentbit-hhmzn                1/1     Running     0               2m9s
cattle-logging-system       rancher-logging-kube-audit-fluentbit-z5tk5                1/1     Running     0               2m9s
cattle-logging-system       rancher-logging-kube-audit-fluentd-0                      2/2     Running     0               2m11s
cattle-logging-system       rancher-logging-kube-audit-fluentd-configcheck-ac2d4553   0/1     Completed   0               2m30s
cattle-logging-system       rancher-logging-rke2-journald-aggregator-dgtdx            1/1     Running     0               2m37s
cattle-logging-system       rancher-logging-rke2-journald-aggregator-gqcq7            1/1     Running     0               2m37s
cattle-logging-system       rancher-logging-rke2-journald-aggregator-mp764            1/1     Running     0               2m37s
cattle-logging-system       rancher-logging-root-fluentbit-2xj55                      1/1     Running     0               2m13s
cattle-logging-system       rancher-logging-root-fluentbit-qvnjd                      1/1     Running     0               2m14s
cattle-logging-system       rancher-logging-root-fluentbit-rcc9w                      1/1     Running     0               2m13s
cattle-logging-system       rancher-logging-root-fluentd-0                            2/2     Running     0               2m16s
cattle-logging-system       rancher-logging-root-fluentd-configcheck-ac2d4553         0/1     Completed   0               2m28s
cattle-monitoring-system    alertmanager-rancher-monitoring-alertmanager-0            2/2     Running     1 (12m ago)     12m
cattle-monitoring-system    helm-install-rancher-monitoring-dpksl                     0/1     Completed   0               13m
cattle-monitoring-system    prometheus-rancher-monitoring-prometheus-0                3/3     Running     0               12m
cattle-monitoring-system    rancher-monitoring-grafana-7fc56c48dc-5hqlg               4/4     Running     0               12m
cattle-monitoring-system    rancher-monitoring-kube-state-metrics-65cd94fffc-gf7mq    1/1     Running     0               12m
cattle-monitoring-system    rancher-monitoring-operator-56f6964d8f-npczs              1/1     Running     0               12m
cattle-monitoring-system    rancher-monitoring-prometheus-adapter-786df96896-6c4x9    1/1     Running     0               12m
cattle-monitoring-system    rancher-monitoring-prometheus-node-exporter-8hqrl         1/1     Running     0               12m
cattle-monitoring-system    rancher-monitoring-prometheus-node-exporter-mnf9d         1/1     Running     0               12m
cattle-monitoring-system    rancher-monitoring-prometheus-node-exporter-prs2f         1/1     Running     0               12m
```

**Comment by TachunLin** [comment]
Verified **failed** on `v1.2.0-rc2`. Move the issue to implement stage for further investigation. 

After we enabled the `monitoring` addon and check the Harvester dashboard
The `Cluster Metrics` and `VM metrics` both displayed with **page not found**

  ![image](https://github.com/harvester/harvester/assets/29251855/0be9552b-0853-4de6-af8b-d058c15e7775)
  
  ![image](https://github.com/harvester/harvester/assets/29251855/8a6f0fb6-a405-462b-afa6-2119635d4f83)
  
* Detailed video clips  


https://github.com/harvester/harvester/assets/29251855/4bfdc161-9078-4364-8a5c-89fc359d86be

* Refresh page but still don't work


https://github.com/harvester/harvester/assets/29251855/c20cff0d-6b79-47cf-ac52-eef5cac096ac

#### Support Bundle
[supportbundle_a8808085-22a6-4e6c-86a2-43ca478f5535_2023-06-26T07-56-50Z.zip](https://github.com/harvester/harvester/files/11865802/supportbundle_a8808085-22a6-4e6c-86a2-43ca478f5535_2023-06-26T07-56-50Z.zip)

### Additional context
Add the lablel of topology.kubernetes.io/zone=zone1 to zone3 to each node before enabling the addons.

**Comment by w13915984028** [comment]
@TachunLin The grafana 404 issue was tracked in https://github.com/harvester/harvester/issues/4087, and solved in master-head,  `v1.2.0-rc2` did not include that fix yet.

**Comment by TachunLin** [comment]
Thanks @w13915984028 for the reminder, let me prepare another Harvester cluster with latest v1.2-head to continue the validation.

**Comment by TachunLin** [comment]
Verified **passed** all the fresh install Harvester test plans list  https://github.com/harvester/harvester/issues/2992#issuecomment-1592505587 on `master-7487edde-head`

While running the single node upgrade to verify the monitoring and logging addon. 
I found after successfully upgrade complete, the `monitoring` and `loggging` addon are still **enabled**. 

https://github.com/harvester/harvester/assets/29251855/de28fc02-1fbe-4c87-94df-32bfcab6b31f


@w13915984028 May I confirm what should be the expected status of monitoring and logging addon after upgrading from v1.1.2 to v1.2.0. Should we keep the addon status to `off` or `enabled` ?

**Comment by w13915984028** [comment]
@TachunLin 

When upgrading from v1.1.2 -> v1.2.0, the `rancher-monitoring` and `rancher-logging` are converted from `managedchart` to `addon`, but they are kept as `enabled`, this is for consistent user experience.  Moreover, e.g. the users may have configured `clusterrule` and `clusterflow` to collect logs, by nature, they will expect those features are still there after upgrading. Thanks.

**Comment by TachunLin** [comment]
Verified **passed** the single node upgrade test plans list https://github.com/harvester/harvester/issues/2992#issuecomment-1592505587 on master-7487edde-head

For the multiple nodes upgrade test plans, since the 3 nodes upgrade test encountered similar failure track in issue 
#4095 [[BUG] Upgrade stuck on post-drain job with the unresponsive VM](https://github.com/harvester/harvester/issues/4095#top)

Thus we move to ready for testing and verify the multi nodes upgrade test plan once it's more stable in upcoming releases.

**Comment by TachunLin** [comment]
Testing the four nodes upgrade to v1.2.0-rc3, check the monitoring and logging addon feature of the following 

- [x] Multi nodes: 
     - [x]  Monitoring and logging addons keep enabled
     - [x]  Can disable monitoring and logging addon
     - [x]  Monitoring pods created and running correctly on `cattle-monitoring-system` namespace
     - [x]  Logging pods created and running correctly on `cattle-logging-system` namespace 
     - [x]  Monitoring chart display correctly on cluster and VM metrics
     - [x]  (Cluster)Flow/(Cluster)Output work as expected

* But found the addon state after successfully, the `rancher-logging` and `rancher-monitoring` display in **processing** state  
![image](https://github.com/harvester/harvester/assets/29251855/ae9702ce-6ca7-4747-b6b4-2ee03f0b14c8)

* Check the single node after upgrade complete in comparison:
  ![image](https://github.com/harvester/harvester/assets/29251855/3081e838-ae13-4d94-8dee-df4ffdf2ff85)



#### Additional context
Harvester logs: 
[supportbundle_11d2f3d9-aa72-4014-8259-4ae064435c19_2023-07-11T14-41-08Z.zip](https://github.com/harvester/harvester/files/12016059/supportbundle_11d2f3d9-aa72-4014-8259-4ae064435c19_2023-07-11T14-41-08Z.zip)


@w13915984028, may I confirm whether above state would be the expected behavior.

**Comment by TachunLin** [comment]
When we attempt to upgrade the 3 nodes and 4 nodes Harvester from v1.1.2 to v1.2.0-rc4 to validate the multiple cluster monitoring and logging addons. 

For the 4 nodes upgrade, we encountered issue: #4285 
For the 3 nodes upgrade, we encounter issue #4290 

Each of them stuck in upgrade and failed to complete the upgrade process. 

Thus I suggest moving the issue to ready for testing and verify again the upgrade test scenario on the later releases.

**Comment by TachunLin** [comment]
Verified again on the successfully upgrade from v1.1.2 to v1.2.0-rc4. 

* After upgrade to v1.2.0-rc4, we can find the Monitoring and Logging addon now display `DeploySuccessful` instead of `Processing` 
  ![image](https://github.com/harvester/harvester/assets/29251855/2794bff0-25b6-41e4-af79-5628d58e319b)

* The monitoring chart and vm metrics display as expected 
  ![image](https://github.com/harvester/harvester/assets/29251855/d1f26c70-2b87-4bcc-9898-13c29cc57714)

* We can disable the monitoring and logging addon, and all related pod will be removed. 
  ![image](https://github.com/harvester/harvester/assets/29251855/52517db6-fb1a-497e-9e41-33158fe7f37e)

**Comment by TachunLin** [comment]
Found an minor issue about the state update while disabling the addon.

* When we disable the monitoring addon, it would keep in `disabling` can't change to `disabled`
![image](https://github.com/harvester/harvester/assets/29251855/52c02dff-dd0d-4016-a2a7-a65443095d48)

* Unless we refresh the page, it can display the correct state
![image](https://github.com/harvester/harvester/assets/29251855/945ca692-3e91-4e26-a447-f212547e19b8)

* Same thing also happen while disabling the logging addon
![image](https://github.com/harvester/harvester/assets/29251855/5bb3e2b9-0ac4-4094-bc56-087efee02f0d)

@w13915984028, 
Do you think we can fix this minor issue in v1.2.0 release or suggest we use a separate issue to track. Thank you.

**Comment by w13915984028** [comment]
@TachunLin

The addon status syncing is tracked in https://github.com/harvester/harvester/issues/4125, the latest UI should have fixed this issue.  cc @DaiYuzeng 

At the meantime, there is a side-effect issue https://github.com/harvester/harvester/issues/4306, which is being solved.

thanks.

**Comment by TachunLin** [comment]
Thanks @w13915984028 for checking the issue. 
I think we can monitor the status syncing and details page as stated in above issues. 

Verified **fixed** on `v1.2.0-rc4`. Close this issue.

### Result
We have passed all test plans designed in https://github.com/harvester/harvester/issues/2992#issuecomment-1592505587 for fresh install and upgrade scenarios

#### Fresh install 
1. - [x] Single and multiple nodes (Enable) : 
2. - [x] Single and multiple nodes (Disable) : 

#### Upgrade to v1.2.0-rc4
1. - [x] Single and multiple nodes (Enable) : 
2. - [x] Single and multiple nodes (Disable) : 


### Test Information
* Test Environment: 3 nodes harvester on bare machines
* Harvester version: v1.2.0-rc1, v1.2.0-rc4

### Verify Steps
  1. <details><summary>Fresh install </summary>
        
        1. Open Advanced -> Addons
        2. Check the `rancher-logging` and `rancher-monitoring` are disabled
        3. ssh to Harvester management node
        4. Execute `kubectl get pods` 
        5. Check no logging and monitoring pods exists 
        6. Create several virtual machines
        7. Enable the logging and monitoring addon
        8. Check the addons display `DeploySuccessful` 
        9. Check the related pods of logging and monitoring created and in running state 
        10. Check the prometheus monitoring chart on node page 
        11. Check the vm metrics on nodes and vm page 
        12. Check the cpu and memory usage of node 
        13. Deploy an elastic search service, refer to https://docs.harvesterhci.io/v1.1/logging/harvester-logging#example-outputs
        14. Configure (cluster Flow and cluster Output in Logging page
        15. Check the logging feature work as expected and can search Harvester log from elastic search service  
       
 </details> 


  2. <details><summary>Upgrade test </summary>
        
        1. Prepare a multiple nodes v1.1.2 Harvester cluster 
        2. Confirm the logging and monitoring service works before upgrade
        3. Upgrade to v1.2.0-rc4
        4. Open Advanced -> Addons
        5. Check the `rancher-logging` and `rancher-monitoring` remain `DeploySuccessful`
        6. ssh to Harvester management node
        7. Execute `kubectl get pods` 
        8. Check the related pods of logging and monitoring created and in running state 
        9. Create several virtual machines
        10. Check the prometheus monitoring chart on node page 
        11. Check the vm metrics on nodes and vm page 
        12. Check the cpu and memory usage of node 
        13. Deploy an elastic search service, refer to https://docs.harvesterhci.io/v1.1/logging/harvester-logging#example-outputs
        14. Configure (cluster Flow and cluster Output in Logging page
        15. Check the logging feature work as expected and can search Harvester log from elastic search service  
        16. Disable the `rancher-logging` and `rancher-monitoring` on Addons page
        17. Check the `rancher-logging` and `rancher-monitoring` can be disable and in correct state
        18. ssh to Harvester management node
        19. Check the no related pods of logging and monitoring pods exists
       
 </details>
